active-rfcs/0000-build-performance.md

```diff
@@ -0,0 +1,187 @@
+
+This change will make this behavior defined and enforced. You can achieve dynamic usage by making the attribute static, but dynamically render a hydrated component like so:
+
+```astro
+---
+import Clock from '../components/Clock.jsx';
+const { shouldHydrate } = Astro.props;
+---
+{ shouldHydrate ? <Clock client:idle /> : <Clock /> }
+```
+
+Astro will be able to build the above correctly, but only JavaScript will be added to the page if the `shouldHydrate` condition is truthy.
+
+## Local directive
+
+In order to replace `Astro.resolve`, which can't be used to build things like styles and images, this RFC proposes a new sugar syntax:
+
+```astro
+<img local:src="../images/penguin.png" />
+```
+
+Note that this can already be achieved without sugar syntax by using an import statement:
+
+```astro
+---
+const imgUrl = '../images/penguin.png';
+---
+
+<img src={imgUrl} />
+```
+
+The sugar syntax is being introduced because we feel the above is harder to learn and less obvious of a solution.
+
+## Deprecation of Astro.resolve
+
+The `Astro.resolve` method was added to enable adding relative links to assets in the `src/` folder. However, since this is a function it can take dynamic values that are not compatible with a static build. For example:
+
+```astro
+---
+const { animal } = Astro.props;
+---
+<img src={Astro.resolve(`../images/${animal}.png`)} />
+```
+
+Instead, you can use `import.meta.glob` to dynamically add a URL:
+
+```astro
+---
+const images = import.meta.glob('../images/*.png');
+
+const { animal } = Astro.props;
+
+const animalUrl = Object.entries(images)
+.filter(([path]) => (
+  path.includes(animal)
+))
+.map(([, mod]) => mod.default)[0];
+---
+
+<img src={animalUrl} />
+```
+
+The above will result in *all* of the images in `../images/` getting built, but only the one you select will be used.
+
+# Motivation
+
+- Astro is currently only able to build sites with a few hundred pages. Since the introduction of `getStaticPaths` we have known that developers would want to build site into the thousands or tens of thousands of pages.
+- Astro's build process relies on scanning the rendered HTML and then *update* the HTML as well, to replace assets with the hashed paths in the build.
+- Because of the above, performance has actually regressed in __0.21__, even though it was never the best even before.
+- In order to support SSR in the future we have to move away from page-scanning as the way to find and build assets, since SSR apps by their nature *cannot* be rendered ahead of time.
+
+# Detailed design
+
+## Enforced static use of client directives
+
+The client directives such as `client:load`, `client:idle`, etc will need to be defined in the hydrated component where they are used, and not rendered dynamically. For example the following is __not allowed__:
+
+```astro
+---
```

Comment:
nit: might be worth putting `// not allowed!` in the code snippet below. I know you have it above but I have to admit that i missed it while scanning the diff and began to read the code snippet